manager/integration/tests/test_scheduling.py

```diff
@@ -694,6 +694,91 @@ def test_replica_auto_balance_node_best_effort(client, volume_name):  # NOQA
+    """
+    Scenario: replica auto-balance should not cause rebuild loop.
+              - replica auto-balance set to `best-effort`
+              - volume data locality set to `best-effort`
+              - volume has 1 replica
+
+    Issue: https://github.com/longhorn/longhorn/issues/4761
+
+    Given no existing volume in the cluster.
+    And set `replica-auto-balance` to `best-effort`.
+    And create a volume:
+        - set data locality to `best-effort`
+        - 1 replica
+
+    When attach the volume to self-node.
+    And wait for the volume to be healthy.
+    Then the only volume replica should be already on the self-node or
+         get rebuilt one time onto the self-node.
+    And volume have 1 replica only and it should be on the self-node.
+         - check 15 times with 1 second wait interval
+
+    When repeat the test for 10 times.
+    Then should pass.
+    """
+    for i in range(10):
+        replica_auto_balance_with_data_locality_test(
+            client, f'{volume_name}-{i}'
+        )
+
+
+def replica_auto_balance_with_data_locality_test(client, volume_name):  # NOQA
+    common.cleanup_all_volumes(client)
+
+    common.update_setting(client,
+                          SETTING_REPLICA_AUTO_BALANCE, "best-effort")
+
+    self_node = get_self_host_id()
+    number_of_replicas = 1
+    volume = client.create_volume(name=volume_name,
+                                  size=str(200 * Mi),
+                                  numberOfReplicas=number_of_replicas,
+                                  dataLocality="best-effort")
+    volume = common.wait_for_volume_detached(client, volume_name)
+
+    volume.attach(hostId=self_node)
+    volume = wait_for_volume_healthy(client, volume_name)
+
+    for _ in range(30):
+        volume = client.by_id_volume(volume_name)
+        if len(volume.replicas) == number_of_replicas and \
+                volume.replicas[0]['hostId'] == self_node:
+            break
+        try:
+            if len(volume.replicas) == number_of_replicas + 1:
+                is_rebuilded = True
+            assert len(volume.replicas) == number_of_replicas
+            assert volume.replicas[0]['hostId'] == self_node
+            assert volume[VOLUME_FIELD_ROBUSTNESS] == VOLUME_ROBUSTNESS_HEALTHY
+            break
+        except AssertionError:
+            # Longhorn should clean up the redundant replica correctly, so
+            # breaking this for loop asserts the result of a single rebuild.
+            if is_rebuilded and len(volume.replicas) == number_of_replicas:
+                break
+            time.sleep(RETRY_INTERVAL)
+
+    assert len(volume.replicas) == number_of_replicas, \
+        f"Unexpected replica count for volume {volume_name}.\n"
+    assert volume.replicas[0]['hostId'] == self_node, \
+        f"Unexpected replica host ID for volume {volume_name}.\n"
+    assert volume[VOLUME_FIELD_ROBUSTNESS] == VOLUME_ROBUSTNESS_HEALTHY
```

Comment:
Why not just check these 10 times? I don't really understand the repeated check above? Could you explain more for range(30) and range(15)? I feel nothing different.

Reply:
The range(30) interaction allows time for the data locality to rebuild the replica.
The range(15) interaction asserts there is no more rebuilding because after the first rebuilding the replica should be on the correct node.
The caller range(10) repeats the tests to ensure the original issue can be reproduced on every run of this e2e case. This is because though this case has high reproducibility, there is a possibility we miss it on a single run.

Reply:
After discussing with @innobead , we will skip this case from the daily run.